reasoner/Reasoner.java

```diff
@@ -73,41 +78,31 @@ public Reasoner(ConceptManager conceptMgr, LogicManager logicMgr,
         this.conceptMgr = conceptMgr;
         this.traversalEng = traversalEng;
         this.logicMgr = logicMgr;
-        Context.Query defaultContext = new Context.Query(context, new Options.Query());
-        defaultContext.producer(Either.first(EXHAUSTIVE));
-        this.controllerRegistry = new ControllerRegistry(actor(), traversalEng, conceptMgr, logicMgr, defaultContext);
+        this.controllerRegistry = new ControllerRegistry(actor(), traversalEng, conceptMgr, logicMgr, context);
         this.explainablesManager = new ExplainablesManager();
     }
 
     public ControllerRegistry controllerRegistry() {
         return controllerRegistry;
     }
 
-    private boolean mayReason(Disjunction disjunction, Context.Query context) {
-        if (!context.options().infer() || context.transactionType().isWrite() || !logicMgr.rules().hasNext()) {
-            return false;
-        }
-        return mayReason(disjunction);
-    }
-
-    private boolean mayReason(Disjunction disjunction) {
-        for (Conjunction conj : disjunction.conjunctions()) {
-            Set<Variable> vars = conj.variables();
-            List<Negation> negs = conj.negations();
-            if (iterate(vars).flatMap(v -> iterate(v.inferredTypes())).distinct().anyMatch(this::hasRule)) return true;
-            if (!negs.isEmpty() && iterate(negs).anyMatch(n -> mayReason(n.disjunction()))) return true;
+    public FunctionalIterator<? extends ConceptMap> execute(Disjunction disjunction, TypeQLMatch.Modifiers modifiers, Context.Query context) {
```

Comment:
`execute()` is actually the main entry point into this class, so we move it to the top and do all the main switching logic at the top